core/src/test/java/org/infinispan/conflict/impl/CrashedNodeDuringConflictResolutionTest.java

```diff
@@ -0,0 +1,201 @@
+public class CrashedNodeDuringConflictResolutionTest extends BaseMergePolicyTest {
+
+   private static final Log log = LogFactory.getLog(CrashedNodeDuringConflictResolutionTest.class);
+   private static boolean trace = log.isTraceEnabled();
+
+   private static final String PARTITION_0_VAL = "A";
+   private static final String PARTITION_1_VAL = "B";
+   private static final String BEFORE_CR_CRASH_KEY = "BEFORE_CR_CRASH";
+   private static final String DURING_CR_CRASH_KEY = "DURING_CR_CRASH";
+   private static final String AFTER_CR_RESTART_KEY = "AFTER_CR_CRASH";
+   private static final String RESOLVED_VALUE = "RESOLVED";
+   private static final String[] ALL_KEYS = new String[] {BEFORE_CR_CRASH_KEY, DURING_CR_CRASH_KEY, AFTER_CR_RESTART_KEY};
+   private static final EntryMergePolicy POLICY = (preferredEntry, otherEntries) -> {
+      Object key = preferredEntry != null ? preferredEntry.getKey() : ((CacheEntry)otherEntries.get(0)).getKey();
+      return new ImmortalCacheEntry(key, RESOLVED_VALUE);
+   };
+   private static final KeyPartitioner PARTITIONER = new TestKeyPartioner();
+
+   public CrashedNodeDuringConflictResolutionTest() {
+      super(DIST_SYNC, null, new int[]{0, 1}, new int[]{2, 3});
+      this.mergePolicy = POLICY;
+      this.valueAfterMerge = RESOLVED_VALUE;
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder dcc = cacheConfiguration();
+      dcc.clustering()
+            .cacheMode(cacheMode).partitionHandling().whenSplit(partitionHandling).mergePolicy(mergePolicy)
+            .hash().keyPartitioner(PARTITIONER);
+      createClusteredCaches(numMembersInCluster, dcc, new TransportFlags().withFD(true).withMerge(true));
+      waitForClusterToForm();
+   }
+
+   @Override
+   protected void beforeSplit() {
+      // Ignore
+   }
+
+   @Override
+   protected void duringSplit(AdvancedCache preferredPartitionCache, AdvancedCache otherCache) {
+      for (String key : ALL_KEYS) {
+         cache(p0.node(0)).put(key, PARTITION_0_VAL);
+         cache(p1.node(0)).put(key, PARTITION_1_VAL);
+      }
+
+      for (String key : ALL_KEYS) {
+         assertCacheGet(key, PARTITION_0_VAL, p0.getNodes());
+         assertCacheGet(key, PARTITION_1_VAL, p1.getNodes());
+      }
+   }
+
+   @Override
+   protected void performMerge() {
+      CompletableFuture<StateRequestCommand> future = createStateRequestFuture()
+            .whenComplete((requestCommand, throwable) -> {
+               if (trace) log.trace("crashCacheManager(2)");
+               TestingUtil.crashCacheManagers(manager(2));
+            });
+
+      for (String key : ALL_KEYS) {
+         assertCacheGet(key, PARTITION_0_VAL, p0.getNodes());
+         assertCacheGet(key, PARTITION_1_VAL, p1.getNodes());
+      }
+
+      fork(() -> partition(0).merge(partition(1), false));
+
+      try {
+         future.get(60, TimeUnit.SECONDS);
```

Comment:
The point I was trying to make about refactoring `CompleteFutureOnStateRequestHandler` is that we should do as much as possible on the test thread, so we have better control over when the node is killed. So I had something more like this in mind:

      partition(0).merge(partition(1), false);
      blockedStateRequest.get(10, TimeUnit.SECONDS);

      if (trace) log.trace("crashCacheManager(2)");
      TestingUtil.crashCacheManagers(manager(2));